ethereum/contracts/ShutdownSwitch.sol

```diff
@@ -0,0 +1,181 @@
+ * Inheriting from `ShutdownSwitch` will allow a bridge contract to use the isEnabled modifier to prevent transactions
+ * when the shutdown switch is engaged via a guardian voting mechanism.
+ *
+ * This contract provides a public function that allows the guardians to vote to disable / enable transfers.
+ * When a predetermined number of guardians vote to disable, then transfers are blocked until the number of
+ * disable votes is reduced below that threshold.
+ *
+ * NOTE: The number of votes to enable is irrelevant. As long as there are enough votes to disable, transfers are blocked.
+ *
+ * An important design goal of this implementation is to minimize the impact (gas) to users, so the check to determine
+ * if transfers are enabled is a simple boolean check.
+ *
+ * The shutdown status is only updated when a vote is cast. When that happens, this contract gets the current guardian set.
+ * It validates that the vote is for an active guardian. If so, it updates their vote, counts the number of
+ * votes to disable, and updates the shutdown status.
+ *
+ * Votes are cast using an authorization proof that is unique to each guardian and the wallet public key used to generate
+ * the proof. That wallet should be different from the guardian's private key. When a guardian casts a vote, it must be sent
+ * from that wallet.
+ *
+ * The authorization proof is generated using the wallet public key and the guardian private key. When a vote is cast,
+ * the code uses the authorization proof and the message sender (wallet public key) to recover the guardian public key.
+ * That is compared to the current list of active guardians, and if a match is found, then the vote is valid.
+ *
+ * NOTE: The status is updated when any active guardian casts a vote, whether it changes anything or not. This means that,
+ * after a guardian set update, the shutdown status can be updated by casting an enable vote when you are already enabled.
+ *
+ * This contract provides public getters to query the shutdown status. It also emits events when valid votes are cast,
+ * or the shutdown status changes.
+ */
+
+import "./libraries/external/BytesLib.sol";
+import "./interfaces/IWormhole.sol";
+
+abstract contract ShutdownSwitch {
+    using BytesLib for bytes;
+
+    /// @dev A map of all guardians that have active votes to disable. If a guardian is removed from the guardian set while they
+    /// have an active vote to disable, they would get left in the map. The current assumption is that this will have minimal impact,
+    /// because they would never be referenced again anyway.
+    mapping(address => bool) private shutdownVotes;
+
+    bool private enabled = true;
+
+    /// @dev Returns the number of votes required to disable transfers.
+    function requiredVotesToShutdown() public view returns (uint16) {
+        return computeRequiredVotesToShutdown(getCurrentGuardianSet().keys.length);
+    }
+
+    /// @dev Returns the current number of votes to disable transfers.
```

Comment:
The comments between 53 and 58 feel a bit too similar.  I wonder if we could differentiate a bit more between the to to drive clarity?

Reply:
Added more detail.